Segfault on building TSAN-enabled Julia

I am unable to build a TSAN-enabled Julia on any recent version of Julia on Ubuntu 20.04 with an x86_64 architecture. However, I *can* build an ASAN-enabled version; just not a TSAN-enabled version (recommended by @vtjnash for debugging #47957).

I am building inside a dockerfile for reproducibility:

```dockerfile
FROM ubuntu:20.04

ENV DEBIAN_FRONTEND=noninteractive

RUN apt-get update && apt-get install -y --no-install-recommends \
    build-essential \
    cmake \
    git \
    wget \
    ca-certificates \
    curl \
    gpg-agent \
    software-properties-common \
    python3 \
    python3-dev \
    tar \
    xz-utils \
    gfortran

RUN wget https://apt.llvm.org/llvm.sh && chmod +x llvm.sh && ./llvm.sh 13

ARG JLVERSION=1.8.3

WORKDIR /downloads

RUN git clone https://github.com/JuliaLang/julia && \
    cd julia && \
    git checkout v${JLVERSION}

# Build:
RUN mkdir -p /workspace && cd julia && contrib/tsan/build.sh /workspace
```
With this, one builds a TSAN-enabled Julia, version 1.8.4, with `docker build --build-arg JLVERSION=1.8.4 -t julia .`. On 1.8.3 and 1.8.4, I see the following segfault during the build:

```
    JULIA /workspace/tsan/usr/lib/julia/corecompiler.ji
Segmentation fault (core dumped)
make[1]: *** [/downloads/julia/sysimage.mk:61: /workspace/tsan/usr/lib/julia/corecompiler.ji] Error 139
make: *** [/downloads/julia/Makefile:82: julia-sysimg-ji] Error 2
The command '/bin/sh -c mkdir -p /workspace && cd julia && contrib/tsan/build.sh /workspace' returned a non-zero code: 2
```

I am unable to build older versions due to this issue: https://github.com/JuliaLang/julia/issues/43896#issuecomment-1366960579.

Here are the last 100 lines of my compilation attempt in case this helps. I tried it on 1.9.0-beta2:

```
    #25 jl_parse_eval_all toplevel.c:? (libjulia-internal-debug.so.1+0xe9c7a)
    #26 ijl_load_ ??:? (libjulia-internal-debug.so.1+0xe93fe)
    #27 ijl_load ??:? (libjulia-internal-debug.so.1+0xe9fdb)
    #28 _finish_julia_init init.c:? (libjulia-internal-debug.so.1+0xa8437)
    #29 julia_init ??:? (libjulia-internal-debug.so.1+0xa7f78)
    #30 jl_repl_entrypoint ??:? (libjulia-internal-debug.so.1+0x14f442)
    #31 jl_load_repl ??:? (libjulia-debug.so.1+0x12323)
    #32 main ??:? (julia-debug+0x4e5b3d)

SUMMARY: ThreadSanitizer: signal-unsafe call inside of a signal :? in __gnu_cxx::new_allocator<std::_Rb_tree_node<std::pair<unsigned long const, objfileentry_t> > >::deallocate(std::_Rb_tree_node<std::pair<unsigned long const, objfileentry_t> >*, unsigned long)
==================
_start at /workspace/tsan/usr/bin/julia-debug (unknown line)
==================
WARNING: ThreadSanitizer: signal-unsafe call inside of a signal (pid=14878)
    #0 free /workspace/srcdir/llvm-project/compiler-rt/lib/tsan/rtl/tsan_interceptors_posix.cpp:726 (julia-debug+0x462ab0)
    #1 jl_print_native_codeloc :? (libjulia-internal-debug.so.1+0x11dfc8)
    #2 jl_print_bt_entry_codeloc :? (libjulia-internal-debug.so.1+0x11e10a)
    #3 jl_critical_error :? (libjulia-internal-debug.so.1+0x153675)
    #4 sigdie_handler signal-handling.c:? (libjulia-internal-debug.so.1+0x152c7d)
    #5 __tsan::CallUserSignalHandler(__tsan::ThreadState*, bool, bool, int, __sanitizer::__sanitizer_siginfo*, void*) /workspace/srcdir/llvm-project/compiler-rt/lib/tsan/rtl/tsan_interceptors_posix.cpp:2021 (julia-debug+0x4613c8)
    #6 ForwardingMemoryManager::allocateCodeSection(unsigned long, unsigned int, unsigned int, llvm::StringRef) :? (libjulia-codegen-debug.so.1+0x1b877d)
    #7 llvm::RuntimeDyldImpl::emitSection(llvm::object::ObjectFile const&, llvm::object::SectionRef const&, bool) [clone .localalias.14] RuntimeDyld.cpp:? (libLLVM-14jl.so+0x28d5ffd)
    #8 llvm::orc::BasicIRLayerMaterializationUnit::materialize(std::unique_ptr<llvm::orc::MaterializationResponsibility, std::default_delete<llvm::orc::MaterializationResponsibility> >) [clone .localalias.8] Layer.cpp:? (libLLVM-14jl.so+0x2856730)
    #9 _jl_compile_codeinst(_jl_code_instance_t*, _jl_code_info_t*, unsigned long, llvm::orc::ThreadSafeContext) jitlayers.cpp:? (libjulia-codegen-debug.so.1+0x17bf25)
    #10 jl_generate_fptr_for_unspecialized_impl ??:? (libjulia-codegen-debug.so.1+0x17ced8)
    #11 jl_compile_method_internal ??:? (libjulia-internal-debug.so.1+0x5a281)
    #12 _jl_invoke gf.c:? (libjulia-internal-debug.so.1+0x5fc20)
    #13 ijl_invoke ??:? (libjulia-internal-debug.so.1+0x5f9ea)
    #14 jl_toplevel_eval_flex :? (libjulia-internal-debug.so.1+0xe6060)
    #15 jl_parse_eval_all toplevel.c:? (libjulia-internal-debug.so.1+0xe9c7a)
    #16 ijl_load_ ??:? (libjulia-internal-debug.so.1+0xe93fe)
    #17 ijl_load ??:? (libjulia-internal-debug.so.1+0xe9fdb)
    #18 _finish_julia_init init.c:? (libjulia-internal-debug.so.1+0xa8437)
    #19 julia_init ??:? (libjulia-internal-debug.so.1+0xa7f78)
    #20 jl_repl_entrypoint ??:? (libjulia-internal-debug.so.1+0x14f442)
    #21 jl_load_repl ??:? (libjulia-debug.so.1+0x12323)
    #22 main ??:? (julia-debug+0x4e5b3d)

SUMMARY: ThreadSanitizer: signal-unsafe call inside of a signal :? in jl_print_native_codeloc
==================
==================
WARNING: ThreadSanitizer: signal-unsafe call inside of a signal (pid=14878)
    #0 free /workspace/srcdir/llvm-project/compiler-rt/lib/tsan/rtl/tsan_interceptors_posix.cpp:726 (julia-debug+0x462ab0)
    #1 jl_print_native_codeloc :? (libjulia-internal-debug.so.1+0x11dfd1)
    #2 jl_print_bt_entry_codeloc :? (libjulia-internal-debug.so.1+0x11e10a)
    #3 jl_critical_error :? (libjulia-internal-debug.so.1+0x153675)
    #4 sigdie_handler signal-handling.c:? (libjulia-internal-debug.so.1+0x152c7d)
    #5 __tsan::CallUserSignalHandler(__tsan::ThreadState*, bool, bool, int, __sanitizer::__sanitizer_siginfo*, void*) /workspace/srcdir/llvm-project/compiler-rt/lib/tsan/rtl/tsan_interceptors_posix.cpp:2021 (julia-debug+0x4613c8)
    #6 ForwardingMemoryManager::allocateCodeSection(unsigned long, unsigned int, unsigned int, llvm::StringRef) :? (libjulia-codegen-debug.so.1+0x1b877d)
    #7 llvm::RuntimeDyldImpl::emitSection(llvm::object::ObjectFile const&, llvm::object::SectionRef const&, bool) [clone .localalias.14] RuntimeDyld.cpp:? (libLLVM-14jl.so+0x28d5ffd)
    #8 llvm::orc::BasicIRLayerMaterializationUnit::materialize(std::unique_ptr<llvm::orc::MaterializationResponsibility, std::default_delete<llvm::orc::MaterializationResponsibility> >) [clone .localalias.8] Layer.cpp:? (libLLVM-14jl.so+0x2856730)
    #9 _jl_compile_codeinst(_jl_code_instance_t*, _jl_code_info_t*, unsigned long, llvm::orc::ThreadSafeContext) jitlayers.cpp:? (libjulia-codegen-debug.so.1+0x17bf25)
    #10 jl_generate_fptr_for_unspecialized_impl ??:? (libjulia-codegen-debug.so.1+0x17ced8)
    #11 jl_compile_method_internal ??:? (libjulia-internal-debug.so.1+0x5a281)
    #12 _jl_invoke gf.c:? (libjulia-internal-debug.so.1+0x5fc20)
    #13 ijl_invoke ??:? (libjulia-internal-debug.so.1+0x5f9ea)
    #14 jl_toplevel_eval_flex :? (libjulia-internal-debug.so.1+0xe6060)
    #15 jl_parse_eval_all toplevel.c:? (libjulia-internal-debug.so.1+0xe9c7a)
    #16 ijl_load_ ??:? (libjulia-internal-debug.so.1+0xe93fe)
    #17 ijl_load ??:? (libjulia-internal-debug.so.1+0xe9fdb)
    #18 _finish_julia_init init.c:? (libjulia-internal-debug.so.1+0xa8437)
    #19 julia_init ??:? (libjulia-internal-debug.so.1+0xa7f78)
    #20 jl_repl_entrypoint ??:? (libjulia-internal-debug.so.1+0x14f442)
    #21 jl_load_repl ??:? (libjulia-debug.so.1+0x12323)
    #22 main ??:? (julia-debug+0x4e5b3d)

SUMMARY: ThreadSanitizer: signal-unsafe call inside of a signal :? in jl_print_native_codeloc
==================
==================
WARNING: ThreadSanitizer: signal-unsafe call inside of a signal (pid=14878)
    #0 free /workspace/srcdir/llvm-project/compiler-rt/lib/tsan/rtl/tsan_interceptors_posix.cpp:726 (julia-debug+0x462ab0)
    #1 jl_print_native_codeloc :? (libjulia-internal-debug.so.1+0x11dff6)
    #2 jl_print_bt_entry_codeloc :? (libjulia-internal-debug.so.1+0x11e10a)
    #3 jl_critical_error :? (libjulia-internal-debug.so.1+0x153675)
    #4 sigdie_handler signal-handling.c:? (libjulia-internal-debug.so.1+0x152c7d)
    #5 __tsan::CallUserSignalHandler(__tsan::ThreadState*, bool, bool, int, __sanitizer::__sanitizer_siginfo*, void*) /workspace/srcdir/llvm-project/compiler-rt/lib/tsan/rtl/tsan_interceptors_posix.cpp:2021 (julia-debug+0x4613c8)
    #6 ForwardingMemoryManager::allocateCodeSection(unsigned long, unsigned int, unsigned int, llvm::StringRef) :? (libjulia-codegen-debug.so.1+0x1b877d)
    #7 llvm::RuntimeDyldImpl::emitSection(llvm::object::ObjectFile const&, llvm::object::SectionRef const&, bool) [clone .localalias.14] RuntimeDyld.cpp:? (libLLVM-14jl.so+0x28d5ffd)
    #8 llvm::orc::BasicIRLayerMaterializationUnit::materialize(std::unique_ptr<llvm::orc::MaterializationResponsibility, std::default_delete<llvm::orc::MaterializationResponsibility> >) [clone .localalias.8] Layer.cpp:? (libLLVM-14jl.so+0x2856730)
    #9 _jl_compile_codeinst(_jl_code_instance_t*, _jl_code_info_t*, unsigned long, llvm::orc::ThreadSafeContext) jitlayers.cpp:? (libjulia-codegen-debug.so.1+0x17bf25)
    #10 jl_generate_fptr_for_unspecialized_impl ??:? (libjulia-codegen-debug.so.1+0x17ced8)
    #11 jl_compile_method_internal ??:? (libjulia-internal-debug.so.1+0x5a281)
    #12 _jl_invoke gf.c:? (libjulia-internal-debug.so.1+0x5fc20)
    #13 ijl_invoke ??:? (libjulia-internal-debug.so.1+0x5f9ea)
    #14 jl_toplevel_eval_flex :? (libjulia-internal-debug.so.1+0xe6060)
    #15 jl_parse_eval_all toplevel.c:? (libjulia-internal-debug.so.1+0xe9c7a)
    #16 ijl_load_ ??:? (libjulia-internal-debug.so.1+0xe93fe)
    #17 ijl_load ??:? (libjulia-internal-debug.so.1+0xe9fdb)
    #18 _finish_julia_init init.c:? (libjulia-internal-debug.so.1+0xa8437)
    #19 julia_init ??:? (libjulia-internal-debug.so.1+0xa7f78)
    #20 jl_repl_entrypoint ??:? (libjulia-internal-debug.so.1+0x14f442)
    #21 jl_load_repl ??:? (libjulia-debug.so.1+0x12323)
    #22 main ??:? (julia-debug+0x4e5b3d)

SUMMARY: ThreadSanitizer: signal-unsafe call inside of a signal :? in jl_print_native_codeloc
==================
Allocations: 0 (Pool: 0; Big: 0); GC: 0
Aborted (core dumped)
make[1]: *** [/downloads/julia/sysimage.mk:61: /workspace/tsan/usr/lib/julia/corecompiler.ji] Error 134
make: *** [/downloads/julia/Makefile:84: julia-sysimg-ji] Error 2
```

The total compilation output resulted in 398 MB of similar ThreadSanitizer warnings being dumped to stdout.

I think so too.

Also... it still actually seems to segfault before it's completely done the build, so this issue isn't fixed yet. (Took a long time to build everything else).

Now it seems to segfault when it's generating the REPL precompile statements:

```
    JULIA usr/lib/julia/sys-o.a
Generating REPL precompile statements... 40/40
ERROR: LoadError: failed process: Process(`/downloads/julia/usr/bin/julia -O0 --trace-compile=/tmp/jl_tI9Idk --sysimage /downloads/julia/usr/lib/julia/sys.ji --cpu-target=native --startup-file=no -i --color=yes -e 'import REPL; REPL.Terminals.is_precompiling[] = true'`, ProcessExited(66)) [66]

Stacktrace:
 [1] top-level scope
   @ /downloads/julia/contrib/generate_precompile.jl:10
in expression starting at /downloads/julia/contrib/generate_precompile.jl:7
*** This error is usually fixed by running `make clean`. If the error persists, try `make cleanall`. ***
make[1]: *** [sysimage.mk:89: /downloads/julia/usr/lib/julia/sys-o.a] Error 1
make: *** [Makefile:88: julia-sysimg-release] Error 2
```

---

By the way, does anybody know the last working TSAN build? I'm happy to use an older version of Julia (>=1.6), all I wanted to do was some data race debugging for #47957, and now I'm yak shaving... If there's an older version I can build successfully, I'd be very happy.

--- a/contrib/tsan/Make.user.tsan
+++ b/contrib/tsan/Make.user.tsan
@@ -11,6 +11,3 @@ USE_BINARYBUILDER_LLVM=1
 
 override SANITIZE=1
 override SANITIZE_THREAD=1
-
-# default to a debug build for better line number reporting
-override JULIA_BUILD_MODE=debug

